### DISCUSSION TITLE
Mutex feature question?

### DISCUSSION THREAD

**Discussion by DrewEaster** [jobrunr/jobrunr#124] [Q&A] [answered]
Hi there,

Awesome project you've created here!

Just wanted to confirm my understanding that there's no mechanism in the free version of JobRunr that would allow some kind of conditional scheduling logic across multiple jobs? Is the mutex feature from the pro version the only way to do this?

We have to implement jobs that hit an external, rate-limited API. The rate limits are tenant wide, and it means it would be feasible for us to have multiple jobs that are essentially exhausting the same rate limit. Thus, we want to ensure such jobs would not be executed at the same time. The mutex feature seems like a good fit for this, but just want to confirm there's no other feature in JobRunr we could lean on to achieve this.

Thanks!

**Comment by rdehuyss** [comment, marked as answer]
Hmmm, it depends: what is the rate limit of the external API? 
I first thought the mutex would indeed solve this but it is limited to 1. You could use more mutexes but that may not be an optimal solution.

Perhaps I can add something like throttling/rate limiting as it feels differently as I think about it.

Can I also ask what StorageProvider you are using?

**Reply by DrewEaster** [reply]
Hey, thanks for responding!

To give a bit more detail. We're hitting an external API to sync data to our system. That API limits API calls at 60 per minute, per tenant. The issue is that we have scenarios where it's feasible that we'd have two (or more) sync jobs we'd want to schedule that would be of fetching data for the same tenant. So, whilst it would be feasible to implement rate limiting within the logic of a single job, if multiple jobs were running for the same tenant, the rate limit for that tenant would still be exceeded. So, what we'd be after is a way to ensure that we could schedule multiple jobs for the same tenant, but trust that JobRunr would prevent them running in parallel.

We've not actually started using JobRunr yet, but would be using Postgres (RDBMS) as a storage provider assuming we do go ahead.

**Reply by DrewEaster** [reply]
A bit more detail...we're syncing data on behalf of our users (so the user is authenticating with the external API using their own credentials, giving us permission to ingest their data from that system into ours). So, the tenant in our scenario is one of our users. So, the rate limit applies on a per user basis, which means we could schedule many jobs for different tenants/users without hitting the 60 per minute limit. We just need control over scheduling where we have multiple jobs being scheduled for one user.

**Reply by rdehuyss** [reply]
Thanks for the extra context.

One more question to make sure that I can give a correct answer: what trigger will schedule the jobs for these tenants? Is it an external event that will enqueue jobs or is it a JobRunr recurring scheduled task that will poll the api every x time (e.g. every 1 minute)?

**Reply by DrewEaster** [reply]
Good question. It could actually be both those :-) The likelihood is that, once a sync is setup, we'll want to schedule a recurring check to see if there's any new data to ingest. However, it's also likely that we'd want to give a user the chance to manually trigger the same check where they may know the remote data has changed and thus wish to expedite the reingestion rather than wait for the next scheduled recurring check.

**Reply by DrewEaster** [reply]
The recurring check would not be that frequent. At most only once a day.

**Reply by rdehuyss** [reply]
I see - so if indeed the biggest risk is the parallel consumption of the external API, then the mutex feature definitely should cover your concerns.

**Reply by DrewEaster** [reply]
Ok, great, thanks for your help in clarifying this.
